agnpy/radiative_process/time_evolution.py

```diff
@@ -0,0 +1,247 @@
+import logging
+from typing import Iterable, Callable, Union, Sequence
+
+import numpy as np
+import astropy.units as u
+from astropy.units import Quantity
+from astropy.constants import c, e
+from numpy._typing import NDArray
+
+from agnpy import Blob, InterpolatedDistribution, Synchrotron, SynchrotronSelfCompton
+from agnpy.utils.conversion import mec2
+
+log = logging.getLogger(__name__)
+
+# A function type that takes a Quantity-array (unitless Lorentz gamma factors)
+# and returns a new Quantity-array (with units of energy/time, same length).
+EnergyChangeFnType = Callable[[Quantity], Quantity]
+
+def synchrotron_loss(sync: Synchrotron) -> EnergyChangeFnType:
+    return lambda gamma: sync.electron_energy_loss_rate(gamma) * -1
+
+def ssc_loss(sync: SynchrotronSelfCompton) -> EnergyChangeFnType:
+    return lambda gamma: sync.electron_energy_loss_rate(gamma) * -1
+
+def ssc_thomson_limit_loss(sync: SynchrotronSelfCompton) -> EnergyChangeFnType:
+    return lambda gamma: sync.electron_energy_loss_rate_thomson(gamma) * -1
+
+def fermi_acceleration(blob):
+    return lambda gamma: blob.xi * blob.B_cgs * c * e.gauss
+
+class TimeEvolution:
+    """
+    Evaluates the change of electron spectral density inside the blob.
+
+    Parameters
+    ----------
+    blob : ~agnpy.Blob
+        The blob for which the time evolution will be performed. As a result of the time evolution,
+        the blob.n_e will be replaced with the InterpolatedDistribution.
+    time : ~astropy.units.Quantity
+         Total time for the calculation
+    energy_change_functions : a function, or an array of functions, of EnergyChangeFnType type
+         The function(s) to be used for calculation of energy change rate per gamma values
+         (for energy gain processes, function should return positive values; for loss, negative values)
+    """
+
+    def __init__(self, blob: Blob, time: Quantity, energy_change_functions: Union[EnergyChangeFnType, Sequence[EnergyChangeFnType]]):
+        self._blob = blob
+        self._time = time
+        self._energy_change_functions = energy_change_functions
+
+    def eval_with_fixed_intervals(self, intervals_count=1, method="heun", max_change_per_interval=0.01):
+        """
+        Performs the time evolution of the electron distribution inside the blob, repeating the calculation in the
+        equal-length time intervals.
+
+        Parameters
+        ----------
+        intervals_count : int
+            the calculation will be performed in the N steps of equal duration (time/N),
+            calculating energy change for every gamma value in each step
+            if set to 0 (default), then calculation will use the automatic time splitting, and recalculating energy change rate
```

Comment:
something is not right here, the default is 1 not 0, and below you raise an exception if intervals_count<=0

Reply:
yes, also a leftover from the earlier version, removed